src/Symfony/Component/Encryption/Provider/PhpseclibEncryption.php

```diff
@@ -0,0 +1,183 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Encryption\Provider;
+
+use phpseclib\Crypt\AES;
+use phpseclib\Crypt\Random;
+use phpseclib\Crypt\RSA;
+use Symfony\Component\Encryption\AsymmetricEncryptionInterface;
+use Symfony\Component\Encryption\Exception\DecryptionException;
+use Symfony\Component\Encryption\Exception\EncryptionException;
+use Symfony\Component\Encryption\Exception\InvalidArgumentException;
+use Symfony\Component\Encryption\Exception\SignatureVerificationRequiredException;
+use Symfony\Component\Encryption\Exception\UnableToVerifySignatureException;
+use Symfony\Component\Encryption\Exception\UnsupportedAlgorithmException;
+use Symfony\Component\Encryption\JWE;
+use Symfony\Component\Encryption\SymmetricEncryptionInterface;
+
+if (!class_exists(RSA::class)) {
+    throw new \LogicException('You cannot use "Symfony\Component\Security\Core\Encryption\PhpseclibEncryption" as the "phpseclib/phpseclib:2.x" package is not installed. Try running "composer require phpseclib/phpseclib:^2".');
+}
+
+/**
+ * The secret key length should be 32 bytes, but other sizes are accepted.
+ *
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ *
+ * @experimental in 5.3
+ */
+class PhpseclibEncryption implements SymmetricEncryptionInterface, AsymmetricEncryptionInterface
+{
+    private $secret;
+
+    /**
+     * @var string application secret
+     */
+    public function __construct(string $secret)
+    {
+        $this->secret = $secret;
```

Comment:
Sound's weird to have this compulsory parameter which is not needed for asymmetric crytography.

I get your point that symmetric is more probably used with local secret, while asymmetric public key is more probably provided by end-user.
But this is not always the case (see SodiumVault class) and asymmetric private key is more probably kept local.

IMHO the API should be identical for all kind of cryptography. Either by providing a key in the constructor or in the method.

Reply:
I dont think it is a good idea to provide private/public keys in the constructor for asymmetric encryption. But I get your point. But that means I need to modify the SymmetricEncryptionInterface to accept a second parameter on the `encrypt()` method. 

The consequences is that the asymmetric implementation cannot make a difference between `SymmetricEncryptionInterface::encrypt($message, $key)` and `AsymmetricEncryptionInterface::encrypt($message, $public, $private = null)`. Ie, I need to rename the methods. 

This is all fine, it is currently a bit of a hack =)

Do you have a suggestion for new method names?

Reply:
What's about `encrypt($message, Key $key)`
```php
SymmetricKey extends Key;
AsymmetricKey extends Key;
DoubleAsymmetricKey extends AsymmetricKey;
```

Reply:
Thank you. It took a day for this idea to sink in. 

I've updated the PR.